[5.2] Use \Doctrine\DBAL\Driver\PDOConnection if available

mzur opened the pull request:
This fixes #14356.

Short summary of the problem: I wasn't able to change a column of a database table that already had a `json` or `jsonb` column. Doctrine/DBAL threw an exception "Unknown database type json requested ..." although my database supported JSON just fine.

I found that relatively new database types like `json` are only supported by DBAL if it is able to [detect the version](https://github.com/doctrine/dbal/blob/61e1f860a304c3fb6961ecbec36927b5f20b71c6/lib/Doctrine/DBAL/Connection.php#L430-L435) of the database in use. For example only the PostgreSQL92Platform [supports](https://github.com/doctrine/dbal/blob/61e1f860a304c3fb6961ecbec36927b5f20b71c6/lib/Doctrine/DBAL/Platforms/PostgreSQL92Platform.php#L73) `json` but the "default" PostgreSqlPlatform (which is used when no version can be detected) does not.

DBAL uses the [PDOConnection](https://github.com/doctrine/dbal/blob/61e1f860a304c3fb6961ecbec36927b5f20b71c6/lib/Doctrine/DBAL/Driver/PDOConnection.php) class as a wrapper for PDO to determine if it can check the database versions. But the Laravel database connector always used just PDO, so DBAL was never able to check the database versions and things like the `json` type were not supported.

This PR modifies the database connector to use the PDOConnection if Doctrine/DBAL is installed and PDO if it is not.

GrahamCampbell commented on src/Illuminate/Database/Connectors/Connector.php:
```diff
@@ -84,6 +85,23 @@ public function setDefaultOptions(array $options)
+    public function getPdoClass()
+    {
+        if (class_exists('Doctrine\DBAL\Driver\PDOConnection')) {
+            return \Doctrine\DBAL\Driver\PDOConnection::class;
```
Please import this

mzur replied:
But DBAL may not be installed. Doesn't it throw an error if the class does not exist?

GrahamCampbell replied:
nope, it works just fine

GrahamCampbell replied:
php makes no attempt verify the existence ou autoload ::class definitions.

GrahamCampbell commented on src/Illuminate/Database/Connectors/Connector.php:
```diff
@@ -84,6 +85,23 @@ public function setDefaultOptions(array $options)
+     */
+    public function getPdoClass()
+    {
+        if (class_exists('Doctrine\DBAL\Driver\PDOConnection')) {
```
Please use class notation here.

GrahamCampbell commented on src/Illuminate/Database/Connectors/Connector.php:
```diff
@@ -84,6 +85,23 @@ public function setDefaultOptions(array $options)
+     *
+     * @return string
+     */
+    public function getPdoClass()
```
Why is this public?

mzur replied:
Good point. I wanted to add a test for this first but couldn't make one up.

mzur committed:
Use \Doctrine\DBAL\Driver\PDOConnection if available

If doctrine/dbal is installed the PDOConnection class provided by
the package should be used instead of just PDO. Else e.g. detecting
the database version doesn't work.

GrahamCampbell commented:
I think this would be best targeted at L5.3, which is getting release pretty soon anyway.

oceanwap commented:
Is it also fixed on 5.1 ?

mzur commented:
No, this is only fixed for 5.3+.
